### ISSUE TITLE
IDMapper much slower in 2.8.0 than before

### ISSUE THREAD

**Issue by zhuchcn** [OpenMS/OpenMS#6072]
Hi, we found that IDMapper takes at least twice as much time as before on MS3 data. Not sure if it's faster on MS2. One of the reason might be that in the consensusXML, the specturm IDs of ConsensusFeature are MS3 , but in the idXML, the spectrum IDa of PeptideIdentification are MS2, so they never match. And then it falls back to using RT and MZ for mapping. But still not sure why this is even slower than before.

Another question is, if 2.7.0 is faster, is it OK to still using that with MS3 data?

**Comment by jpfeuffer** [comment]
IdMapper currently uses an extremely inefficient quadratic lookup. We recently added the matching based on scan ID which I think is definitely necessary for a correct mapping (and therefore strongly recommend 2.8).
Of course this adds a quadratic amount of string comparisons plus metavalue lookups.

I think it would benefit from a larger rewrite. How much are we talking in absolute numbers? We did not care so much, since we thought it never is the bottleneck.

**Comment by jpfeuffer** [comment]
Oh wow! No it was always quadratic but only float comparisons. Could honestly be the extra scan id comparison now.
We make sure that the scan id of the precursor scans are matched, so that should be ok and that is why I recommend 2.8 for this.

A non quadratic lookup should also not be hard. We just thought it would never be the bottleneck.

**Comment by zhuchcn** [comment]
Does it use the mzML spectra file to find the precursors? I didn't provide that when running IDMapper and could that be the reason why it's so slow?

**Comment by jpfeuffer** [comment]
Yes but that already happens in IsobaricQuantifier and the feature is annotated with the id_scan_id and with a (quant) scan_id.
In IDMapper it should in theory just be a lookup nrFeats*nrScans IIRC.

**Comment by zhuchcn** [comment]
Got it! So the id_scan_id is written by IsobaricAnalyzer. That makes sense. One more question. May not be relevant to this run time issue. Our samples are fractionated, so we searched library on each fraction, and merged the idXML for FDR, and we ran IsobaricAnalyzer on each fraction, and merge them with FileMerger. My question is, does IDMapper differentiate the fractions from the merged idXML & consensusXML? Because there could be spectra with the same ID from different fractions. We tried to reindex the mzML files at the very beginning to generate non-redundant scan id across all fractions, and just wondering if that's necessary and if that's a good practice. The runtime is the same, btw, 3-4 hours in either way.

**Comment by jpfeuffer** [comment]
I'm afraid it does not do that. Although the information should be there if you enabled annotate_file_origin during ID merging. ConsensusXML knows by the map ID and the ConsensusHeaders.
Middle-term I wanted to include the IDMapping into IsobaricQuantifier directly.

Two more alternatives:
- try the new IDsplitter functionality after FDR, map each ConsensusXML then merge them.
- do a two step FDR approach. A 5-10% FDR on each file. Then the desired FDR on the final consensusXML

**Comment by zhuchcn** [comment]
Those are helpful! For the first one, we want to run Percolator, so that seems to require IDMerger to have all IDRuns merged (with -merge_proteins_add_PSMs). Will it then still be able to split by IDSplitter? I noticed that the merge ID is still kept by IDMapper, but it may require us to keep track on the mapping from initial file name to merge ID. 

What about the way we are doing that makes the spectrum ID all unique in all mzMLs at the beginning? So fraction 1 will be "spectrum=1" to "spectrum=10000", fraction2 will be "spectrum=10001" to "spectrum=2000".

**Comment by jpfeuffer** [comment]
So yes the functionality is relatively new. It should however know from the spectra_files metavalue array and the merge_id, which file it came from.

Your approach with the scan ids seems to be fine though.

**Comment by jpfeuffer** [comment]
Oh btw, I think the tool is weirdly called IDRipper 😅

**Comment by zhuchcn** [comment]
No worry I got you! IDRipper sounds like a cool name! 😄

**Comment by timosachsenberg** [comment]
might be fixed
